### ISSUE TITLE
Doctest is not able to compile on OSX

### ISSUE THREAD

**Issue by arnavb** [onqtam/doctest#126]
## Description

I am trying to get my Travis CI build working with doctest (I'm switching from Catch to this for performance), but my build is not working on OSX (It is for Linux). In my [Travis CI logs](https://travis-ci.org/arnavb/cpp14-project-template/jobs/367163159), I see the following error:
```
Undefined symbols for architecture x86_64:
  "std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::operator<<<char, std::__1::char_traits<char>, std::__1::allocator<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)", referenced from:
      doctest::String doctest::detail::stringifyBinaryExpr<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, char [6]>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const*, char const (&) [6]) in hello_world_test.cpp.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [test/test_runner] Error 1
make[1]: *** [test/CMakeFiles/test_runner.dir/all] Error 2
make: *** [all] Error 2
```

### Steps to reproduce

I'm using CMake, and my invocation command is:

```
cmake -DCMAKE_BUILD_TYPE=$BUILD_TYPE -DCMAKE_CXX_COMPILER=$COMPILER -DBUILD_TESTS=ON ..
```
(`COMPILER` is set to `clang++` and `BUILD_TYPE` is set to `Debug` and `Release`, both of which fail)

The CMake configuration I am using is (for flags):

```CMake
if (CMAKE_CXX_COMPILER_ID MATCHES "Clang" OR CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wextra -Werror")
    set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -g")
    set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -O2")
elseif (CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
    # ...
endif(CMAKE_CXX_COMPILER_ID MATCHES "Clang" OR CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
```

And the test configuration:

```CMake
set(TEST_SOURCES
    factorial_test.cpp
    hello_world_test.cpp
)

add_library(doctest INTERFACE)
target_include_directories(doctest INTERFACE 
    "${CMAKE_SOURCE_DIR}/third_party/doctest" # Contains single header
)

add_executable(test_runner test_runner.cpp ${TEST_SOURCES})
target_link_libraries(test_runner Project-Name-lib doctest)

add_test(all_tests test_runner)
```

### Extra information

* doctest version: **v1.2.8**
* Operating System: **Travis CI OSX XCode image 9.2**
* Compiler+version: **AppleClang 9.0.0.9000039**

**Comment by onqtam** [comment]
weird... I have an OSX build of doctest on travis with the same version of clang:
https://travis-ci.org/onqtam/doctest/jobs/364548718

I also tried your branch locally on Windows with VS 2017 and GCC 7.2 and its all fine - seems an OSX issue only.

Normally I try to resolve serious issues as fast as possible but currently I have no time to spare - I have to prepare 2 full lectures - after 10 days everything will be over and I'll be able to perhaps try this on a Mac - sorry for the inconvenience.

**Comment by arnavb** [comment]
@onqtam Thanks for the quick reply. I'll keep trying to fix this ... somehow. For now, I'm going to isolate this issue by remove all extraneous details. Thanks for that note about the `REQUIRE` asserts. Turns out, I actually needed `CHECK` asserts all along (I just found out the difference recently).

**Comment by arnavb** [comment]
@onqtam Ok Update. I've created another repository in which I'm trying to use doctest, and ... it works fine! Now I have to figure out the difference between it and my repository. The repository is here: https://github.com/arnavb/osx-doctest

**Comment by onqtam** [comment]
Hi there,

So I forked and played with your 2 repositories and here is what I tried:
- from the original repo: when I removed any use of ```std::string``` from the asserts the problem disappeared.
- moving from ```xcode9.2``` to ```xcode9.3``` or ```xcode9``` didn't change anything.
- not setting the CXX standard - didn't affect anything.
- when I moved the implementation of doctest to a separate file however the problem occurred.
- and after that I included ```<iostream>``` in the file with the asserts and ```std::string``` usage in them and the problem went away........................

My bet is that this is a toolchain issue - and it saddens me that such problems occur in 2018. I guess I'll have to investigate this further...

**Comment by arnavb** [comment]
@onqtam Hmmm... that's interesting. I've added a `#include <string>` to `hello_world_test.cpp` and am now waiting for results. Let's see whether this works or not...

**Comment by onqtam** [comment]
I added ```<iostream>``` in that .cpp - not ```<string>``` (which is already included by ```project-abbr/hello_world.hpp```)......

including a header should not result in fixing link errors............. I have exactly 0 pleasant memories of native development under OSX...

**Comment by arnavb** [comment]
@onqtam You're right. I added `<iostream>` and it works! (Finally). However, this solution is ugly on various levels. `<string>` is supposed to have a more complete version of `std::string` since that's what the header is for! Why is `iostream`, which only probably including `<string>` itself, solving a _LINKING PROBLEM_? 😶

**Comment by onqtam** [comment]
finally resolved "properly" - turns out the problem is this: when using clang from xcode - that means libc++. in that case I include ```<iosfwd>``` instead of forward-declaring std::ostream myself. Then if the user includes somewhere only the doctest header and <string> and then does a comparison - then and only then will there be some linker issue which we resolved by including ```<iostream>``` in the tests. This is a toolchain issue when using the ```<iosfwd>``` header - the details are still not clear to me but I managed to get the linker to pull in what is necessary by adding a dummy unused global function in the implementation of doctest that uses operator<< of std::string to print to std::cout.

the ```<iostream>``` header should no longer be necessary - try the version of the header from the ```dev``` branch (soon to be released in master as version 2.0!)

**Comment by agurtovoy** [comment]
FYI, I just ran into this issue with doctest 2.3.1 (macOS, apple-clang 10.0, libc++), with something as simple as this: 

```C++
#include <doctest/doctest.h>
#include <string>

TEST_CASE( "[console] console" ) 
{
    REQUIRE( std::string("hello") == "hello" );
}
```
(see the errors below). Adding explicit `#include <iostream>` helped.

```
Undefined symbols for architecture x86_64:
  "std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::operator<<<char, std::__1::char_traits<char>, std::__1::allocator<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)", referenced from:
      doctest::String doctest::detail::stringifyBinaryExpr<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, char [6]>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const*, char const (&) [6]) in console.test.cpp.o
```

**Comment by onqtam** [comment]
@agurtovoy well that's unfortunate..... are you building with LTO? The only explanation I have is that this function gets optimized out:
https://github.com/onqtam/doctest/blob/2.3.2/doctest/doctest.h#L6009

**Comment by agurtovoy** [comment]
@onqtam Good guess, I was building a release build and had `-Wl,-dead_strip_dylibs` in the args. Just tested a debug configuration, and it does not need the explicit `<iostream>` include.

**Comment by onqtam** [comment]
Gosh... I'll find a way to "fix" it

**Comment by jenisys** [comment]
I have the same issue with newer doctest versions (above 2.2.2).
When I build with `CMAKE_BUILD_TYPE=Release` linking the test executable fails to link due to the missing symbols `doctest::detail::stringifyBinaryExpr(...)`, basically a simple test using assertion macros. 

EXAMPLE: Fails to build test executable
```c++
#include "doctest/doctest.h"
#include <string>
...
```

WORK-AROUND 1: Succeeds to build test executable
```c++
#include "doctest/doctest.h"
#include <string>
#include <sstream>   // NEEDED-FOR: CMAKE_BUILD_TYPE=Release
...
```

WORK-AROUND 2: Succeeds to build test executable
```c++
#define DOCTEST_CONFIG_DISABLE 1
#include "doctest/doctest.h"
#include <string>
...
```

Build details:

```
# -- HINT: Not using -Wl,-dead_strip_dylibs
/Applications/Xcode.app/.../c++ -O3 -DNDEBUG ... 
   -isysroot /Applications/Xcode.app/.../SDKs/MacOSX10.14.sdk 
   -Wl,-search_paths_first -Wl,-headerpad_max_install_names ...
```

NOTE:
The doctest `all_features/assertation_macros.cpp` builds without problems with CMAKE_BUILD_TYPE=Release that uses `-DDOCTEST_CONFIG_DISABLE` while compiling the object.

**Comment by onqtam** [comment]
@jenisys sorry for the late reply. Can you confirm that you are experiencing this with version 2.3.3 (the latest one) as well? That would be really unfortunate...

Using ```DOCTEST_CONFIG_DISABLE``` means that there are no tests and no test runner - it's as if you are not using doctest at all. If even with it you were experiencing these linker errors I would have been even more puzzled.

**Comment by jenisys** [comment]
@onqtam 
Yes, the problem still exists w/ dockest-2.3.3 (and the current HEAD).
Sorry for the late reply.

**Comment by onqtam** [comment]
@drehren I think in your case the problem could be solved by using [```DOCTEST_CONFIG_ASSERTION_PARAMETERS_BY_VALUE```](https://github.com/onqtam/doctest/blob/master/doc/markdown/configuration.md#doctest_config_assertion_parameters_by_value) - can you try and confirm that it indeed solves the issue?

**Comment by drehren** [comment]
Thanks @onqtam ! It did solve the issue.

**Comment by onqtam** [comment]
@drehren did you look in the FAQ? It's not mentioned there but I wonder if I should add an entry about this specific issue in there...

**Comment by drehren** [comment]
@onqtam I did look through it now that I remember... I think an entry there would be great.

**Comment by lorrden** [comment]
I had the same issue, but only when I build with asserts enabled and optimisations. I tried to update doctest to 2.4.4, but that did not help. So I added the <iostream> include trick and that worked.

This was on macOS 10.13 Catalina.

**Comment by blurpy** [comment]
I had the same issue with doctest 2.4.6 building on macOS in release mode.
The workaround of adding include of `<iostream>` in the affected test worked fine.

**Comment by onqtam** [comment]
This has finally been fixed by actually including the right headers with clang & libc++ - already in the `dev` branch.
Closing this - will release an official version soon!
